src/storage/src/source/generator.rs

```diff
@@ -256,6 +284,19 @@ fn render_simple_generator<G: Scope<Timestamp = MzOffset>>(
                     }
                 }
                 Event::Progress(Some(offset)) => {
+                    // If we've reached the requested maximum offset, cease.
+                    if offset >= max_offset {
+                        break;
+                    }
+
+                    // If the offset is within the requested snapshot offset,
+                    // don't downgrade the capability yet. This will ensure that
+                    // all data in the snapshot is emitted as a single atomic
+                    // hunk.
```

Comment:
@petrosagg can you check me on this assertion? I didn't actually trace this through reclocking.

Reply:
That is correct but if you want to be super correct I think we'd have to also fast forward the timestamp of the data to be at least `snapshot_offset`, which behaves more or less like an `as_of`

Reply:
Done, thanks.